implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+    ///   an encrypted payload containing the initiator identity / signature / credentials
+    pub(super) async fn encode_message3(&mut self, payload: Vec<u8>) -> Result<Vec<u8>> {
+        let mut state = self.state.clone();
+        // encrypt s.pubKey
+        let s_pub_key = self.get_public_key(&state.s).await?;
+        let c = self.encrypt_and_hash(&mut state, s_pub_key.data()).await?;
+        let mut message3 = c.to_vec();
+
+        // ck, k = HKDF(ck, DH(s, re), 2)
+        let dh = self.dh(&state.s, &state.re).await?;
+        (state.ck, state.k) = self.hkdf(&mut state, Some(&dh)).await?;
+
+        // encrypt payload
+        let c = self
+            .encrypt_and_hash(&mut state, payload.as_slice())
+            .await?;
+        message3.extend(c);
+
+        self.state = state;
+        Ok(message3)
+    }
+
+    /// Decode the third message sent by the initiator
+    pub(super) async fn decode_message3(&mut self, message: Vec<u8>) -> Result<Vec<u8>> {
+        let mut state = self.state.clone();
+        // decrypt rs key
+        let rs_pub_key = Self::read_message3_encrypted_key(&message)?;
+        state.rs = PublicKey::new(self.hash_and_decrypt(&mut state, rs_pub_key).await?, X25519);
+
+        // ck, k = HKDF(ck, DH(e, rs), 2), n = 0
+        let dh = self.dh(&state.e, &state.rs).await?;
+        (state.ck, state.k) = self.hkdf(&mut state, Some(&dh)).await?;
+
+        // decrypt payload
+        let c = Self::read_message3_payload(&message)?;
+        let payload = self.hash_and_decrypt(&mut state, c).await?;
+        self.state = state;
+        Ok(payload)
+    }
+
+    /// Set the final state of the state machine by creating the encryption / decryption keys
+    /// and return the other party identity
+    pub(super) async fn set_final_state(&mut self, role: Role) -> Result<()> {
+        let mut state = self.state.clone();
+        // k1, k2 = HKDF(ck, zerolen, 2)
+        let (k1, k2) = self.hkdf(&mut state, None).await?;
+        let (encryption_key, decryption_key) = if role.is_initiator() {
+            (k2, k1)
+        } else {
+            (k1, k2)
+        };
+        state.status = Ready(HandshakeKeys {
+            encryption_key,
+            decryption_key,
+        });
+
+        self.state = state;
+        Ok(())
+    }
+
+    /// Return the final results of the handshake if we reached the final state
+    pub(super) fn get_handshake_keys(&self) -> Option<HandshakeKeys> {
+        match self.state.status.clone() {
+            Ready(keys) => Some(keys),
+            _ => None,
+        }
+    }
+}
+
+impl Handshake {
+    /// Create a new handshake initialized with
+    ///   - a static key
+    ///   - an ephemeral key
+    ///   - a payload containing the identity, a signature of the static key and the credential of
+    ///     the current party
+    pub(super) async fn new(vault: Arc<dyn XXVault>) -> Result<Handshake> {
```

Comment:
Do you think it would be better to always pass static key?

Reply:
That would also indicate that the caller is responsible for the life cycle of that key. Which is closer to RAII

Reply:
Good idea. I've move the generation one level up in the call-stack to make it easier to switch to retrieving a more permanent key in the future.

Reply:
Good idea. I've move the generation one level up in the call-stack to make it easier to switch to retrieving a more permanent key in the future.